### PR TITLE
TeddySwap Integration

### PR DESCRIPTION
**Pull request by Sluder**

### CODE REVIEW

File: src/dex/teddyswap.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
+import { LiquidityPool } from './models/liquidity-pool';
+import { BaseDataProvider } from '@providers/data/base-data-provider';
+import { Asset, Token } from './models/asset';
+import { BaseDex } from './base-dex';
+import { DatumParameters, DefinitionConstr, DefinitionField, PayToAddress, RequestConfig, SwapFee, UTxO } from '@app/types';
+import { DefinitionBuilder } from '@app/definition-builder';
+import { AddressType, DatumParameterKey } from '@app/constants';
+import { BaseApi } from '@dex/api/base-api';
+import pool from './definitions/teddyswap/pool';
+import order from './definitions/teddyswap/order';
```
Comment (Sluder) [inline src/dex/teddyswap.ts:10]: Still need these added

File: src/dex/teddyswap.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            return !assetName?.toLowerCase()?.endsWith('_nft') && !assetName?.toLowerCase()?.endsWith('_identity');
+        });
+
+        // Irrelevant UTxO
+        if (![2, 3].includes(relevantAssets.length)) {
+            return Promise.resolve(undefined);
+        }
+
+        // Could be ADA/X or X/X pool
+        const assetAIndex: number = relevantAssets.length === 3 ? 0 : 1;
```
Comment (pblake81239) [inline src/dex/teddyswap.ts:70]: const assetAIndex: number = relevantAssets.length === 2 ? 0 : 1;
    const assetBIndex: number = relevantAssets.length === 2 ? 1 : 2;

Comment (Sluder) [reply src/dex/teddyswap.ts:70]: Hmm yeah the `relevantAssets` needs to include lovelace as an option as well

File: src/dex/teddyswap.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        return Promise.all(addressPromises).then((liquidityPools: Awaited<LiquidityPool[]>[]) => liquidityPools.flat());
+    }
+
+    public async liquidityPoolFromUtxo(provider: BaseDataProvider, utxo: UTxO): Promise<LiquidityPool | undefined> {
+        if (!utxo.datumHash) {
+            return Promise.resolve(undefined);
+        }
+
+        const relevantAssets = utxo.assetBalances.filter((assetBalance) => {
+            const assetName = (assetBalance.asset as Asset).assetName;
```
Comment (Sluder) [inline src/dex/teddyswap.ts:60]:
```suggestion
            const assetName = assetBalance.asset === 'lovelace' ? 'lovelace' : assetBalance.assetName;
```

**Comment by dpbeaumont** [comment]
Nice! 🚀

File: src/dex/teddyswap.ts (hunk 3 above)
Comment (pblake81239) [reply src/dex/teddyswap.ts]: : assetBalance.asset.assetName;

Is asset not required here?

File: src/dex/teddyswap.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+        // Load additional pool information
+        const lpToken: Asset = utxo.assetBalances.find((assetBalance) => {
+            const assetName = (assetBalance.asset as Asset).assetName;
+            return assetName?.toLowerCase()?.endsWith('_lp');
+        })?.asset as Asset;
+
+        if (lpToken) {
+            liquidityPool.lpToken = lpToken;
+        }
```
Comment (pblake81239) [inline src/dex/teddyswap.ts:107]: else{
      return Promise.resolve(undefined);
    }
    
    Ive added a promise reject on no Lp

File: src/dex/teddyswap.ts (hunk 3 above)
Comment (Sluder) [reply src/dex/teddyswap.ts]: Yes it is! My comment was wrong but the code reflects the correct path now

File: src/definition-builder.ts [hunk 5]
```diff
@@ -100,6 +109,20 @@ export class DefinitionBuilder {
      * Recursively pull parameters from datum using definition template.
      */
     private extractParameters(definedDefinition: DefinitionField, templateDefinition: DefinitionField, foundParameters: DatumParameters = {}): DatumParameters {
+        if (templateDefinition instanceof Function) {
```
Comment (Sluder) [inline src/definition-builder.ts:112]: Adds support for datum mappings to include dynamic pulling/pushing of parameters. One example would be an order datum not including a stake key, which will result in a different datum. Not all DEXs support this atm, but will push a fix after this PR is in

File: src/dex/teddyswap.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            this.buildSwapOrderPayment(swapParameters, {
+                address: this.orderAddress,
+                addressType: AddressType.Contract,
+                assetBalances: [
+                    {
+                        asset: 'lovelace',
+                        quantity: batcherFee?.value + deposit.value,
+                    },
+                ],
+                datum: datumBuilder.getCbor(),
```
Comment (Sluder) [inline src/dex/teddyswap.ts:212]: Inline?

Comment (dpbeaumont) [reply src/dex/teddyswap.ts:212]: transaction.providerData.tx.payToContract(payToAddress.address, { inline: payToAddress.datum as Datum }, payment);

File: src/dex/minswap.ts [hunk 7]
```diff
@@ -210,6 +210,7 @@ export class Minswap extends BaseDex {
                         },
                     ],
                     datum: datumBuilder.getCbor(),
+                    isInlineDatum: false,
```
Comment (dpbeaumont) [inline src/dex/minswap.ts:213]: Nice approach!

**Review by dpbeaumont — APPROVED** [review]
Looks great! 🚀

**Pull request merged**